BXMSDOC-5473-master-kogito: Finalize creating and running content for Kogito.

sterobin opened the pull request:
@MarianMacik , @jomarko , can you provide the QE final verification of this new "Creating and running your first Kogito services" doc? Use the doc previews, not just the PR changes since the PR is only for the last changes.

@ederign , @evacchi , @tiagobento , feel free to provide SME re-review if you'd like, though little has changed since initial reviews.

Note that the "[Enabling the Kogito VSCode extension for BPMN and DMN modeling](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5383/#proc_kogito-vscode-extension)" section may be changing soon based on some updates from @ederign.

Links:
* [JIRA](https://issues.redhat.com/browse/BXMSDOC-5373)
* [Enterprise doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5383/) - **Refresh preview page to ensure latest.**
* [Community doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5383_COMM/) - Same, but conditioned. **Refresh preview page to ensure latest.**

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/creating-running/proc_kogito-running-app.adoc:
```diff
@@ -5,13 +5,13 @@
 After you design the business decisions and processes for your {PRODUCT} service, you can run your Quarkus or Spring Boot application in one of the following modes:
```
@MarianMacik @jomarko , the example now fails when I try to run it due to this error:

```
Fatal error compiling: error: invalid target release: 1.11
```

I checked the pom.xml and see the `maven-compiler` version is 1.11, which I assume is the issue. This is the default setting when you generate the project. I tried changing, removing, whatever, but not luck. Not sure what the issue is. Perhaps a bug in the default project, or something I'm missing.

evacchi replied:
this should be fixed in master

jomarko replied:
it can be fixed if 1.11 is replaced by 11, I think I saw a PR already with that fix.

jomarko replied:
yes, here https://github.com/kiegroup/kogito-runtimes/commit/b912f137773c5f6571592ac7e64ad12af165dd51#diff-eeac56c431e21541d13ae02a623971e4

sterobin replied:
Ack. Did that and fixed. Thanks.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/creating-running/proc_kogito-interacting-app.adoc:
```diff
@@ -1,20 +1,20 @@
-[id='proc_kogito-testing-app']
+[id='proc_kogito-interacting-app']
```
@MarianMacik , @jomarko , I could not test to confirm this section now with the latest version of Kogito due to the run error I get, in my other comment in this PR.

sterobin replied:
@MarianMacik , @jomarko , never mind, got the project to run with that maven-compiler tweak (fix coming in master) and I re-ran and updated the examples and responses. Also noted that users can alternatively use the Swagger UI.

jomarko commented:
### Comment 1
_After you create your Kogito project, you can create or import Business Process Model and Notation (BPMN) 2.0 business processes, Decision Model and Notation (DMN) decision models, Drools Rule Language (DRL) business rules, XLS or XLSX decision tables, Java services, and other assets in the src/main/resources folder of your project._

Not sure I understand **Java services** if it is a java class, shouldn't be in **src/main/java** ?

jomarko commented:
### Comment 2
_Figure 4.3. Example DMN data types_ is obsolete, types can not be imported  VS code.

jomarko commented:
### Comment 3
Very minor typo
_In the left palette, select DMN decision, drag the node to the canvas, double-click the node to name it Is Adult, and link to it from the input node._ should have **DMN Decison**

jomarko commented:
### Comment 4
_Click the upper-left corner of the decision table to set the hit policy to Unique._ Unique is default hit policy. So basically we can omit this step. I am not aware of plan it should change in future, but doesn't harm to keep the step, decide on your own.

jomarko commented:
### Comment 5
Not sure about versioning. We refer to download latest vsix. What can be fine currently, but I build locally vsix based on kiegroup 7.36.0 editors , and found this https://issues.redhat.com/browse/KOGITO-1558. In other words I didn't verify business process creation procedure.

jomarko commented:
### Comment 6
_CHAPTER 6. INTERACTING WITH A RUNNING KOGITO SERVICE_
Example responses are not correct. If we do not update business process, the response will always have two variables, **person** and **isAdult**

```json
{"id":"28ebc1c5-0741-45c7-bcb4-d0e0f75b5d2f","person":{"name":"John Quark","age":20,"adult":false},"isAdult":true}
```

sterobin commented:
@jomarko , thank you! In reply:

1. Java services meaning like these: https://github.com/kiegroup/kogito-examples/tree/stable/jbpm-quarkus-example/src/main/java/org/kie/kogito/examples But yeah, I guess just Java classes. And you're right about confusion about `/resources`. I've just removed mention of "Java services" and left as ", and other assets in the `src/main/resources` folder of your project." (for DRL, DMN, XLS, and the like).
2. Ugh, oops. Sorry, updated everywhere in DMN doc content but forgot this one. Fixed. Good catch!
3. Wow, good eye! Fixed.
4. Ah, okay. I guess will leave for now if only for user awareness of the choices.
5. Now that I got the project working, I re-ran all examples and updated all requests and responses accordingly. The responses are all now similar to this:
```
{"id":"3af806dd-8819-4734-a934-728f4c819682","person":{"name":"John Quark","age":20,"adult":false},"isAdult":true}
```

spolti commented on doc-content/kogito-docs/src/main/asciidoc/dmn/proc_dmn-model-creating.adoc:
```diff
@@ -21,7 +21,7 @@ endif::[]
 If the DRD does not open in the {PRODUCT} DMN designer, ensure that you have installed and enabled the {PRODUCT} https://github.com/kiegroup/kogito-tooling/releases[VSCode extension].
 
-If the {PRODUCT} DMN designer opens only the XML source of the DMN file and displays an error message, review the reported errors and the DMN model file to ensure all DMN elements are correctly defined.
+If the {PRODUCT} DMN designer opens only the XML source of the DMN file and displays an error message, review the reported errors and the DMN model file to ensure that all DMN elements are correctly defined.
```
it is a little bit trick, since it is not easy to understand the xml and find the issue, does someone can give a hint about it?

sterobin replied:
Agreed. If someone has more specific instruction on handling errors, please let me know and I'll include.

spolti commented on doc-content/kogito-docs/src/main/asciidoc/dmn/proc_dmn-logic-defining.adoc:
```diff
@@ -14,7 +14,7 @@ You can use the {PRODUCT} DMN designer in VSCode to define decision logic for yo
 If the DRD does not open in the {PRODUCT} DMN designer, ensure that you have installed and enabled the {PRODUCT} https://github.com/kiegroup/kogito-tooling/releases[VSCode extension].
 
-If the {PRODUCT} DMN designer opens only the XML source of the DMN file and displays an error message, review the reported errors and the DMN model file to ensure all DMN elements are correctly defined.
+If the {PRODUCT} DMN designer opens only the XML source of the DMN file and displays an error message, review the reported errors and the DMN model file to ensure that all DMN elements are correctly defined.
```
why duplicate things?

sterobin replied:
So the user has it where he/she needs it. It seems to happy pretty easily from what I saw, so want to make sure it's addressed.

spolti replied:
I agree with you, but I was talking about to have it write once but re-used on the other places when needed.

sterobin replied:
@spolti , ah, right, as an include snippet of sorts. I agree and would like to, but we are requested not to do that by the CCS tooling team, with the exception of recurring exclamatories, like the Dev Preview note. But even that I haven't "include"-ed, because the verdict is still shaky, so I've just repeated for now.

spolti commented on doc-content/kogito-docs/src/main/asciidoc/creating-running/ref_kogito-app-examples.adoc:
```diff
@@ -2,7 +2,7 @@
 = Example applications with {PRODUCT} services
 
-{PRODUCT} includes several example applications with {PRODUCT} services to help you develop your own applications. The following list describes some of the examples provided with {PRODUCT}. For the full list of examples and instructions for using them, see the https://github.com/kiegroup/kogito-examples[kogito-examples] page in GitHub.
+{PRODUCT} includes many example applications with {PRODUCT} services to help you develop your own applications. The following list describes some of the examples provided with {PRODUCT}. For the full list of examples and instructions for using them, see the https://github.com/kiegroup/kogito-examples[kogito-examples] page in GitHub.
```
maybe it worth to mention the stable branch that will always point to the latest kogito version. (even the stable branch being the the default one)

sterobin replied:
Sure, can do.

sterobin replied:
@spolti , I add this note:
> NOTE: In the https://github.com/kiegroup/kogito-examples[kogito-examples] repository in GitHub, the example applications in the default `stable` branch use the latest version of {PRODUCT}.  

Good with you?

spolti replied:
sounds good, thanks.

spolti commented on doc-content/kogito-docs/src/main/asciidoc/creating-running/proc_kogito-vscode-extension.adoc:
```diff
@@ -15,7 +15,10 @@ Visual Studio Code (VSCode) is the preferred integrated development environment
 After you enable this extension, any `.bpmn2` and `.dmn` files that you open in VSCode are automatically displayed as graphical models.
 
-If the {PRODUCT} BPMN or DMN designers open only the XML source of a BPMN or DMN file and displays an error message, review the reported errors and the model file to ensure all BPMN or DMN elements are correctly defined.
+If the {PRODUCT} BPMN or DMN designers open only the XML source of a BPMN or DMN file and displays an error message, review the reported errors and the model file to ensure that all BPMN or DMN elements are correctly defined.
```
same comment made on dmn section

sterobin replied:
See previous comment.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/creating-running/proc_kogito-vscode-extension.adoc:
```diff
@@ -15,7 +15,10 @@ Visual Studio Code (VSCode) is the preferred integrated development environment
-If the {PRODUCT} BPMN or DMN designers open only the XML source of a BPMN or DMN file and displays an error message, review the reported errors and the model file to ensure all BPMN or DMN elements are correctly defined.
+If the {PRODUCT} BPMN or DMN designers open only the XML source of a BPMN or DMN file and displays an error message, review the reported errors and the model file to ensure that all BPMN or DMN elements are correctly defined.
+
+NOTE: For new BPMN or DMN models, you can also enter `bpmn.new` or `dmn.new` in a Google Chrome browser to design your BPMN or DMN model in the {PRODUCT} online designer. When you finish creating your model, you can click *Download* in the online designer page to import your BPMN or DMN file into your {PRODUCT} project in VSCode.
```
@ederign do we only support Chrome the the online editors?

sterobin replied:
@cristianonicolai , @ederign , nope, apparently not. Just tested. My bad, I thought it was chrome only since Google created the feature. Changed to "in a web browser to ..." Though this will be changed anyway afterward once the tooling hub goes live.

Thanks @cristianonicolai .

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/creating-running/proc_kogito-running-app.adoc:
```diff
@@ -5,13 +5,13 @@
 
 * *Development mode*: For local testing. On Quarkus, development mode also offers live reload of your processes and rules in your running applications for advanced debugging.
-* *JVM mode*: For compatibility with a Java Virtual Machine (JVM).
+* *JVM mode*: For compatibility with a Java virtual machine (JVM).
```
@cristianonicolai , no, it's just "development mode" and "JVM mode" because it's not proper. In other products' docs, they use both back and forth, but at the source, I see that it's not a proper (capitalized) term.

sterobin commented:
@MarianMacik , thank you! In reply:

1. Strange, that was in initial prereqs in the Quarkus content for the same procedure and I saw a docker folder in the example before, but not in the latest version. I've removed that prereq for now.

2. You're right. I thought I saw it was Chrome only because it's a Google feature, but I just tried in Firefox and worked. Fixed.

3. Good point. I've added the following sentence in **bold.** What do you think? @jomarko ?
> After you create your {PRODUCT} project, you can create or import Business Process Model and Notation (BPMN) 2.0 business processes, Decision Model and Notation (DMN) decision models, Drools Rule Language (DRL) business rules, XLS or XLSX decision tables, and other assets in the `src/main/resources` folder of your project. **You can also include Java classes in the `src/main/java` folder of your project that act as Java services or that provide implementations that you call from your business processes or decisions.**

4. Sorry, thought I had already defined DRD. Looks like just there and in screenshot description. Done.

5. Right, makes sense now about the data assignment names. I've since re-run and updated all the API responses in the next sections, so that this is correct now. I've tried to avoid getting into details on the variables, or on the gateway configuration. Just trying to get the user up and running simply and quickly. More details of all of that are found in our upcoming "Developing process services with Kogito" doc, under SME review, and linked at the end of this doc.

6. Regarding the rule unit package: I posed the same question to @evacchi , and what I understood was that it's `unit:org.acme.kogito.PersonRules` because that's the location of the DRL file, but as I think about it again and look at our other rule unit doc content, it should be format `unit:package.unit`, so seems like in fact it should be `unit:org.acme.kogito.model.PersonRules`. 

More importantly, I just ran the example from the doc with DMN and succeeded, then followed the steps in that [rule unit section](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5383/#proc_kogito-designing-app-rule-units) (just add the DRL file, set rule language to DRL, and add the unit as rule flow group) and the build fails when I  use either format of rule unit. So the instructions seem to be broken or missing something now.

@evacchi , do you know what's missing/wrong with that rule unit doc section?

7. Yes, I removed the mentions of native because @spolti pointed out that it will be supported on only the community version of Quarkus for now, and therefore only the "community" version of Kogito (at least as far as docs are communicating).

8. Well, syntactically correct though sure, hard on the ear. The alternative just could be misread: "service for ordering items that exposes REST"  makes it seem at first glance like the items are exposing the REST and like it's a typo. It's really "A process service [that is] for ordering items and that exposes ...". Anyway, pros and cons. This situation comes up a lot.. Would be better to break into two sentences if needed, but then  it breaks the pattern of the list.

sterobin commented:
@MarianMacik , @jomarko , I've implemented your feedback. However, I've updated the examples section and added a new section on running the examples, per feedback from Kris. Refresh the updated [preview link](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5383/#ref_kogito-app-examples_kogito-creating-running) to see.

@krisv Can you review the updated [examples section](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5383/#ref_kogito-app-examples_kogito-creating-running) and let me know if this works for you? It's based on what's on the Kogito site today.

sterobin commented:
@jomarko :

*  bpmn: Ah, okay. Well, that should be updated in the example, imo. I'd prefer to keep it focused on bpmn2 in the docs for consistency and clarity.
* Messy examples: Agreed, most of the examples are like that. They could all use some polishing in the source.
* Stable branch: Yes, already mentioned actually (see Examples section), and the links all direct to the stable branch.
* isAdult: Oops, fixed. Thanks.
* Example file: Sure, will swap out. Thanks!

MarianMacik commented on doc-content/kogito-docs/src/main/asciidoc/creating-running/proc_kogito-designing-app.adoc:
```diff
@@ -2,22 +2,22 @@
 = Designing the application logic for a {PRODUCT} service using DMN and BPMN
 
-After you create your {PRODUCT} project, you can create or import Business Process Model and Notation (BPMN) business processes, Decision Model and Notation (DMN) decision models, Drools Rule Language (DRL) business rules, XLS or XLSX decision tables, Java services, and other assets in the `src/main/resources` folder of your project.
+After you create your {PRODUCT} project, you can create or import Business Process Model and Notation (BPMN) 2.0 business processes, Decision Model and Notation (DMN) decision models, Drools Rule Language (DRL) business rules, XLS or XLSX decision tables, and other assets in the `src/main/resources` folder of your project. You can also include Java classes in the `src/main/java/org/acme` folder o ...
```
You can also include Java classes in the `src/main/java/org/acme`
It's `src/main/java` in general as package can be configured by a user.

sterobin replied:
@MarianMacik Sure. I had it that way before but thought I should be more specific to this particular example that they are running. But you're right, will revert to that. Thanks!

jomarko commented:
@sterobin thank you for your opinion about bpmn vs bpmn2. I created this PR that should unify naming https://github.com/kiegroup/kogito-examples/pull/167
